### PR TITLE
Fixes #77: retry exceptions

### LINKED ISSUE #77: Retry transient HTTP errors
Looks like some transient error--perhaps if this is common/expected cachix can automatically handle retrying internally? Dunno.

```
FatalError {fatalErrorMessage = "FailureResponse (Response {responseStatusCode = Status {statusCode = 500, statusMessage = \"Internal Server Error\"}, responseHeaders = fromList [(\"Server\",\"nginx\"),(\"Date\",\"Tue, 26 Jun 2018 19:45:20 GMT\"),(\"Content-Type\",\"text/plain; charset=utf-8\"),(\"Transfer-Encoding\",\"chunked\"),(\"Connection\",\"keep-alive\")], responseHttpVersion = HTTP/1.1, responseBody = \"Something went wrong\"})"}
```

store path of cachix used, FWIW: `/nix/store/f1h61j4b01ya3rqpqrmh2z6m4haf0rak-cachix-0.1.0.0/bin/cachix` (which is valiid path on `https://cachix.cachix.org`)

### PR DESCRIPTION
**Pull request by domenkozar**
Fixes #77

### CODE REVIEW

File: cachix/src/Cachix/Client/Commands.hs [hunk 1]
```diff
@@ -205,13 +208,18 @@ pushStorePath env name storePath = do
     (cachixBCClient name)
     (Api.NarInfoC storeHash)
   case res of
-    Right NoContent -> return ()
-    Left err | isErr err status404 -> go sk
-             | otherwise -> panic $ show err -- TODO: retry
+    Right NoContent -> return () -- we're done as store path is already in the cache
```
Comment (roberth) [inline cachix/src/Cachix/Client/Commands.hs]: `pass` (`Protolude`)

File: cachix/src/Cachix/Client/Commands.hs [hunk 2]
```diff
@@ -235,15 +243,13 @@ pushStorePath env name storePath = do
...
-        -- TODO: http retry: retry package?
         NoContent <- escalate <=< (`runClientM` newClientEnv) $ Api.createNar
           (cachixBCClient name)
           (contentType (Proxy :: Proxy Api.XNixNar), conduitToStreaming)
         closeStdout
         exitcode <- waitForStreamingProcess cph
         -- TODO: print process stderr?
-        when (exitcode /= ExitSuccess) $
-          panic $ "Failed with " <> show exitcode <> ": " <> show cmd
+        when (exitcode /= ExitSuccess) $ throwM exitcode
```
Comment (roberth) [inline cachix/src/Cachix/Client/Commands.hs]: This is like calling `exitWith`. Wrap this exception in something that is more informative, including which command was run.

File: cachix/src/Cachix/Client/Commands.hs [hunk 3]
```diff
@@ -287,7 +293,10 @@ pushStorePath env name storePath = do
 
 -- Utils
 
--- TODO: should one error abort the whole pushing process? Or do a retry? Or keep going and then fail?
+-- Retry up to 5 times for each store path
+retryPath :: (RetryStatus -> IO a) -> IO a
+retryPath = recoverAll def
```
Comment (roberth) [inline cachix/src/Cachix/Client/Commands.hs]: `recoverAll` is too broad. We don't want to retry on Ctrl+C for example. Use something like [`isSyncException`](https://hackage.haskell.org/package/safe-exceptions-0.1.7.0/docs/Control-Exception-Safe.html#v:isSyncException) with `Control.Retry.recovering`

Comment (roberth) [reply cachix/src/Cachix/Client/Commands.hs]: Ok

Comment (roberth) [reply cachix/src/Cachix/Client/Commands.hs]: Needs an inline comment about skipping asyncs

**Comment by domenkozar** [comment]
Fixed :)

**Pull request merged**